### PR TITLE
SM2 certificate verification

### LINKED ISSUE #6579: about SM2(ECC) cert verify
how to verify SM2(ECC) certificate 
this is my code:

X509 *issuer
X509 *subject
EVP_PKEY *signing_key = X509_get_pubkey(issuer);
int result = X509_verify(x509, signing_key);

but it return -1;
The certificate chain is OK.
I try the code with RSA.It's OK.
But it is error with  SM2.
I have no idea

the version is 1.1.1

### PR DESCRIPTION
**Pull request by randombit**
Fixes #6579

This is on top of #6757 because being able to set the UID field is required to verify a SM2 certificate (it always uses the SM2 default identifier of 1234567812345678). The new commit is 613de33 I will rebase after #6757 is merged.

This is verification only, there is currently no support for generating a new SM2 certificate but that is the next step after #6757 and this PR are merged.

This is a contribution from Ribose Inc

/cc @ronaldtse

### CODE REVIEW

**Comment by t-j-h** [comment]
It makes no sense to hardwire in a single fixed UID in the verify code.

**Comment by randombit** [comment]
I would welcome suggestions on another way to approach the problem.

**Comment by mattcaswell** [comment]
> It makes no sense to hardwire in a single fixed UID in the verify code.

As I understand it this UID is mandated by the standards so it will never be anything different. Am I correct?

**Comment by InfoHunter** [comment]
> As I understand it this UID is mandated by the standards so it will never be anything different. Am I correct?

I just have a quick look on the SM2 signature standards (GM/T 0003.2-2012), the UID is a variable length string. So which standard did you refer to, when you implemented this '12345678123435678' UID? @randombit

**Comment by ronaldtse** [comment]
@mattcaswell @InfoHunter [GM/T 0009-2012](https://riboseinc.github.io/gmt-0009-2012/) Clause 10 specifies the default user id for SM2 algorithm when it is not set.

[GM/T 0015-2012](https://riboseinc.github.io/gmt-0015-2012/index.html), which specifies the certificate format, does not talk about the "user id" for the SM2 algorithm, but specifies that there are no parameters given to SM2 in a certificate. The certificate example at Annex D.2 can only be decoded if the user id "1234567812345678" is used.

**Comment by mattcaswell** [comment]
I wonder whether this UID should be set in the SM2 code automatically if no UID has already been set? Does it ever make sense to do SM2 without a UID?

**Comment by ronaldtse** [comment]
SM2 does require a UID for calculation, so perhaps you're right 😉

**Comment by richsalz** [comment]
I support this going into the initial 1.1.1 release.

**Comment by randombit** [comment]
It honestly is not clear to me if all SM2 certificates are signed using the default SM2 UID string ("123..."). The standards I have been able to find are completely silent on the matter of how this is set. But it turns out in practice at least the live SM2 certificate chain we have  (from #6579) uses that as the identifier.

It seems fine to me to use "123.." as the default ID if not otherwise set. That avoids most of the SM2 specific code in `a_verify.c`. However there still has to be a check for SM2 NID in order to set `EVP_PKEY_set_alias_type` in order to switch to the SM2 algorithm pkey.

I can make that change (using the default ID unless otherwise set) in either this PR or in #6757. I would prefer in this PR since #6757 is already in pending-2nd-review and I don't want to reset the clock there.

**Comment by InfoHunter** [comment]
> It honestly is not clear to me if all SM2 certificates are signed using the default SM2 UID string ("123...").

This exactly the problem. Unfortunately, due to the unclearness of current standards, the answer is also unclear.

SM2 certificates have never been used in any 'real' SSL/TLS protocol. They are used mainly in a Chinese-modified version of TLS 1.1, which I don't think OpenSSL (or any other crypto related software should support). We would like to make SM2 usable in real TLS protocols, then we could use SM2 certificates there. We are putting efforts on this direction. But we don't know the final outcome, it *might* be different for the algorithm and the certificate, say, what UID to use when signing an SM2 certificate...

So I think currently it would be good to adopt 'using the default ID unless otherwise set' method instead of assuming all SM2 certificates are signed using the default 123... ID...

**Comment by ronaldtse** [comment]
I agree with @InfoHunter on:
> So I think currently it would be good to adopt 'using the default ID unless otherwise set' method instead of assuming all SM2 certificates are signed using the default 123... ID...

Could we move this ahead for now, by moving the default UID into general SM2 computation within this PR?

**Comment by t-j-h** [comment]
It need to work off the user being able to set the personalisation string. What concerns me is this use of a default conceptually makes no sense in this context.

**Comment by randombit** [comment]
> It need to work off the user being able to set the personalisation string.

Do you mean, the user needs to be able to set the personalization string during the certificate verification operation?

**Comment by t-j-h** [comment]
-1 blocker for now

**Comment by mattcaswell** [comment]
See #7266

**Comment by mlindner** [comment]
Is there any updates coming on this? @randombit

File: crypto/objects/obj_dat.h [hunk 1]
```diff
@@ -1074,9 +1074,10 @@ static const unsigned char so[7746] = {
     0x2A,0x85,0x03,0x07,0x01,0x02,0x01,0x01,0x02,  /* [ 7718] OBJ_id_tc26_gost_3410_2012_256_paramSetB */
     0x2A,0x85,0x03,0x07,0x01,0x02,0x01,0x01,0x03,  /* [ 7727] OBJ_id_tc26_gost_3410_2012_256_paramSetC */
     0x2A,0x85,0x03,0x07,0x01,0x02,0x01,0x01,0x04,  /* [ 7736] OBJ_id_tc26_gost_3410_2012_256_paramSetD */
+    0x2A,0x81,0x1C,0xCF,0x55,0x01,0x83,0x75,       /* [ 7745] OBJ_sm2_with_SM3 */
```
Comment (mlindner) [inline crypto/objects/obj_dat.h]: Shouldn't this go next to the other SM2/SM3 changes after entry 7331?

**Comment by mlindner** [comment]
General comment: Can we replace the mixed case "sm2_with_SM3" with something like "SM2_with_SM3" or "sm2_with_sm3". This is used in several locations with either underscores or hyphens.

**Comment by randombit** [comment]
Seems like this is not wanted? Closing

**Pull request closed without merging**